File: codejam/2008/round_2/star_wars/solution.hs
{-

xi + yi + zi - pi T <= x + y + z <= xi + yi + zi + pi T
xi + yi - zi - pi T <= x + y - z <= xi + yi - zi + pi T
xi - yi + zi - pi T <= x - y + z <= xi - yi + zi + pi T
xi - yi - zi - pi T <= x - y - z <= xi - yi - zi + pi T

c1 <= x + y + z <= c0
c3 <= x + y - z <= c2
c5 <= x - y + z <= c4
c7 <= x - y - z <= c6

c1 - x <= y + z <= c0 - x
-c6 + x <= y + z <= -c7 + x
c3 - x <= y - z <= c2 - x
-c4 + x <= y - z <= -c5 + x

c1 <= c0
-c6 <= -c7
c3 <= c2
-c4 <= -c5

c1 + c7 <= 2x <= c0 + c6
c3 + c5 <= 2x <= c2 + c4

c1 + c7 <= c0 + c6
c3 + c5 <= c2 + c4
c1 + c7 <= c2 + c4
c3 + c5 <= c0 + c6

-}

import Data.List (intersperse)

infixr 0 <|
(<|) f x = f x
(<||) f g x = f <| g x

data Ship = Ship { x :: Double, y :: Double, z :: Double, p :: Double } deriving (Show)

c'' :: Double -> Int -> Ship -> Double
c'' t i ship
    = x ship
    + ((-1)^(div i 4)) * y ship
    + ((-1)^(div i 2)) * z ship
    + ((-1)^i) * t * p ship

c' :: [Ship] -> Double -> Int -> Double
c' ships t i
    | even i = foldr1 min <| map (c'' t i) ships
    | otherwise = foldr1 max <| map (c'' t i) ships

reachable :: [Ship] -> Double -> Bool
reachable ships t
    = c 0 >= c 1 && c 2 >= c 3 && c 4 >= c 5 && c 6 >= c 7
    && min (c 0 + c 6) (c 2 + c 4) >= max (c 1 + c 7) (c 3 + c 5)
    where c = c' ships t

upperBound :: [Ship] -> Double
upperBound ships = snd
    <| head
    <| dropWhile (not <|| uncurry reachable)
    <| map ((,) ships) (iterate (*2) 1)

binarySearch :: [Ship] -> Int -> Double -> Double -> Double
binarySearch ships p lb ub
    | p == 0 = m
    | reachable ships m = binarySearch ships (p - 1) lb m
    | otherwise = binarySearch ships (p - 1) m ub
    where m = ((lb + ub) / 2)

parse :: String -> Ship
parse s = Ship { x = ds!!0, y = ds!!1, z = ds!!2, p = ds!!3 }
    where ds = map (read :: String -> Double) <| words s

solve :: IO String
solve = do
    n <- readLn :: IO Int
    s <- sequence <| replicate n getLine
    let ships = map parse s
    return
        <| show
        <| binarySearch ships 127 0
        <| upperBound ships

main :: IO ()
main = (readLn :: IO Int)
    >>= sequence <|| flip replicate solve
    >>= putStrLn
        <|| concat
        <|| intersperse "\n"
        <|| map (uncurry (++))
        <|| zip (map (("Case #" ++) <|| (++ ": ") <|| show) [1..])

File: codejam/2008/round_2/star_wars/solution.rs
lib::run!();

struct Ship {
    x: f64,
    y: f64,
    z: f64,
    p: f64,
}

fn read() -> Vec<Ship> {
    lib::input!(usize as n);
    let mut ships: Vec<Ship> = Vec::with_capacity(n);
    for _ in 0..n {
        lib::input!([f64; 4] as [x, y, z, p]);
        ships.push(Ship { x, y, z, p });
    }
    ships
}

fn cc(ships: &[Ship], i: u32, t: f64) -> f64 {
    ships
        .iter()
        .map(|ship| {
            ship.x
                + (-1i64).pow(i >> 2) as f64 * ship.y
                + (-1i64).pow(i >> 1) as f64 * ship.z
                + (-1i64).pow(i) as f64 * t * ship.p
        })
        .fold(f64::NAN, if i & 1 == 0 { f64::min } else { f64::max })
}

fn reachable(ships: &[Ship], t: f64) -> bool {
    let c = |i| cc(ships, i, t);
    f64::min(c(0) + c(6), c(2) + c(4)) >= f64::max(c(1) + c(7), c(3) + c(5))
        && c(0) >= c(1)
        && c(2) >= c(3)
        && c(4) >= c(5)
        && c(6) >= c(7)
}

fn solve(ships: Vec<Ship>) -> String {
    let mut upper_bound = 1f64;
    while !reachable(&ships, upper_bound) {
        upper_bound *= 2f64;
    }
    let mut lower_bound = 0f64;
    for _ in 0..128 {
        let mid_point = (upper_bound + lower_bound) / 2f64;
        if reachable(&ships, mid_point) {
            upper_bound = mid_point;
        } else {
            lower_bound = mid_point;
        }
    }

    format!("{upper_bound}")
}
